[2단계] 블랙잭(딜러)

RoadtoS7 opened the pull request:
2단계 요구사항을 구현이 간단하게 끝날 줄 알았는데 변경된 파일도 많고, 커밋도 많네요..!

수정하다보니 일급 컬렉션을 활용할 수 있는 부분을 최대한 활용해보고자 했습니다!
Dealer와 Player 간의 로직 중복을 피하기 위해 상속을 사용했습니다!
객체의 역할이 어색한 것이 있다면 피드백 부탁드립니다!!
그리고 테스트 코드에서는 디미터 법칙에 어긋나는 경우가 종종 있는데,  테스트에서도 이를 지켜주는 것이 좋은지 궁금합니다! 혹은 상황에 따라 어쩔 수 없이 .을  여러번 찍는 코드도 유닛테스트에서는 사용하는 것이 괜찮은지 궁금합니다!

꼼꼼한 피드백 감사합니다! 이번에도 잘 부탁드립니다!!~ 🤗

RoadtoS7 committed:
refactor: TrumpCard에서 allCards메서드의 파라미터 이름 변경

- 수정 이유
변경된 파라미터 타입의 이름에 따라서 파라미터의 이름도 바꾸어주었다.

RoadtoS7 committed:
refactor: CardScoreRule.idealScore(sumWhereAceOne:sumeWhereAceEleven)메서드 로직 리팩터링

ace를 11점으로 계산한 값이 21점을 넘지 않는다면 최적의 점수가 되기 때문에 이를 반환한다.
만일 21점을 넘는다면, ace를 1점으로 계산한 값이 최적의 점수가 될 수 밖에 없기 때문에 이를 반환하도록 수정했다.

RoadtoS7 committed:
doc: README에 2단계 요구사항 분석 추가

RoadtoS7 committed:
feat: Dealer(딜러)를 Player를 상속받아서 구현

- Dealer도 게임 참가자이기에 Player와 동일한 역할을 함으로 Player를 상속받아 구현했다
- Dealer의 이름이 '딜러'로 지정되어있으므로 이름없이 CardDeck만 입력받아서 객체를 생성할 수 있도록 생성자를 추가로 구현했다

RoadtoS7 committed:
feat: BlackjackAnswerReader가 카드 점수가 16이하인 딜러의 응답을 받을 때 유닛테스트 구현

RoadtoS7 committed:
feat: BlackjackAnswerReader가 카드 점수가 16이하인 딜러의 응답으로 yes를 반환하도록 구현

RoadtoS7 committed:
feat: BlackjackAnswerReader 리팩토링

RoadtoS7 committed:
feat: WinLoseResult에 대한 유닛 테스트 구현

RoadtoS7 committed:
feat: WinLoseResult의 failable intializer 구현

WinLoseResult는 winCount, loseCount로 양수만 가질 수 있다. 만일 intializer로 음수가 들어온다면 nil을 반환한다

RoadtoS7 committed:
refactor: WinLoseResult의 failable intializer 리팩터링

RoadtoS7 committed:
refactor: WinLoseResult가 어떤 Player의 승패결과인지 알 수 있도록 Player을 프로퍼티로 갖도록 수정

RoadtoS7 committed:
refactor: Dealer의 answer() 메서드 제거

- 제거 이유: 사용하지 않음

RoadtoS7 committed:
feat: 점수를 가지고 player의 승패 결과를 계산하는 BlackjackJudge의 유닛테스트 작성

RoadtoS7 committed:
feat: 점수를 가지고 player의 승패 결과를 계산하는 BlackjackJudge 구현

RoadtoS7 committed:
feat: BlackjackJudge.winLoseResult() 리팩토링

RoadtoS7 committed:
feat: main로직에서 dealer도 블랙젝 게임에 참여하도록 수정함

RoadtoS7 committed:
feat: 딜러가 카드를 수령했음을 알리는 OutputView.printDealerReceiveCard 메서드 구현

RoadtoS7 committed:
refactor: BlackjackStateDelegate.afterRecevieCard에서 딜러가 카드를 수령했을 때 메시지 출력하도록 수정

RoadtoS7 committed:
refactor: PlayerNameParser에서 로그 메시지 출력 구문 삭제

RoadtoS7 committed:
feat: WinLoseResult에서 특정 유저의 승패 기록을 꺼내는 subscript에 대한 유닛테스트 구현

RoadtoS7 committed:
feat: 게임 참가자들의 승패결과를 담는 WinLoseResults 구현

RoadtoS7 committed:
feat: 특정 참가자의 점수를 기준으로 참가자들의 승패 결과를 계산하는 BlackjackJudge.winLoseResults() 유닛 테스트 구현

RoadtoS7 committed:
feat: 특정 참가자의 점수를 기준으로 참가자들의 승패 결과를 계산하는 BlackjackJudge.winLoseResults() 구현

RoadtoS7 committed:
feat: BlackjackJudge.winLoseResults(with:comparingWith:) 리팩터링

RoadtoS7 committed:
refaactor: BlackjackJudge.winLoseResults의 with 파라미터를 of 로 네이밍 변경

RoadtoS7 committed:
fix: WinLoseResults의 subscript 구현이 이전 커밋에 포함되지 않아서 발생하는 이슈 해결

RoadtoS7 committed:
feat: OutputView에서 최종 승패를 출력하는 메서드 구현

RoadtoS7 committed:
feat: WinLoseResult를 출력물로 만들어주는 WinLoseResultFormatter 구현

RoadtoS7 committed:
refactor: main에 승패결과를 계산해서 출력하는 로직 추가

RoadtoS7 committed:
refactor: 출력할 때 줄 간격 수정

RoadtoS7 committed:
feat: BlackjackGameJudge에서 딜러의 점수가 21점을 넘어가면 나머지 참가자들이 무조건 우승하는 유닛테스트 구현

RoadtoS7 committed:
refactor: BlackjackGameJudge.winLoseResult()에서 승패수를 구할 때 딜러의 점수가 21점을 넘으면 무조건 참가자들이 우승하는 것으로 수정

RoadtoS7 committed:
refactor: BlackjackGameJudge.winLoseResult() 리팩토링

RoadtoS7 committed:
refactor: 모든 forammter를 Util 디렉토리로 이동

RoadtoS7 committed:
refactor: WinLoseFormatter를 WinLoseResultFormatter로 네이밍 변경

RoadtoS7 committed:
feat: BlackjackScore의 유닛테스트 구현

RoadtoS7 committed:
feat: BlackjackScores.playerScores의 유닛 테스트 구현

- playerScores는 딜러를 제외한 참가자들의 점수만 리턴한다.

RoadtoS7 committed:
feat: BlackjackScores.playerScores 로직 구현

RoadtoS7 committed:
feat: BlackjackScorese.dealerScore 구현

RoadtoS7 committed:
feat: BlackjackGame.start()의 결과로 BlackjackScores를 반환받도록 구현

RoadtoS7 committed:
refactor: WinLoseResults에서 subsciprt를 제거하고, player점수와 딜러의 점수를 프로퍼티로 구분해서 갖도록 수정

zdodev commented on Blackjack/Dealer.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+class Dealer: Player {
```
`딜러` 라는 이름을 가지면서 플레이어 역할을 할 수 있도록 상속해보셨네요. 좋습니다!👍

zdodev commented on Blackjack/BlackjackGameJudge.swift:
```diff
@@ -0,0 +1,52 @@
+            partialResult + result.loseCount
+        })
+        
+        let counterpartResult = WinLoseResult(player: counterpart.player, winCount: totalLoseCount, loseCount: totalWinCount)!
```
`winCount` 에 `totalLoseCount` 를 전달하고 있는데, 전달 이름이 어색해보입니다. 의도하신 부분일까요?

RoadtoS7 replied:
딜러의 loseCount는 나머지 게임 참가자들의 winCount 총합과 같아서 이렇게 표현했었습니다!
근데 주신 피드백을 바탕으로 생각해보니 변수 이름으로 용도나 역할을 담는게 더 가독성이 좋다고 생각되네요!

[71c1588](https://github.com/next-step/swift-blackjack/pull/12/commits/71c1588b5d7b70ccedce4d742fac059e7afac3f2)  커밋에서 주신 피드백을 바탕으로 BlackjackGameJudge에 존재하는 변수 명들을 수정해보았습니다! 피드백 감사합니다!!

zdodev commented on Blackjack/Player.swift:
```diff
@@ -24,3 +24,13 @@ class Player {
     }
 }
+
+extension Player: Hashable {
```
`Hashable` 을 채택하신 이유가 있을까요?

RoadtoS7 replied:
`Player` 를 프로퍼티로 갖는 `BlakcjackScore`를 `Comparable` 프로토콜을 따르도록 만들려고 하니 `Player`가 `Hashable` 프로토콜을 채택하도록 만들어야 했습니다!

zdodev replied:
`Comparable` 프로토콜을 채택하고 비교 가능한 프로퍼티를 만들기 위해서 `Hashable` 프로토콜이 아니라 `Equatable` 프로토콜만 따르면 되지 않을까 싶었습니다!
`Hashable` 프로토콜이 `Equatable` 프로토콜을 상속하고 있어서 문제는 없지만 현재 코드 상에서 `Hashable` 한 기능을 사용하고 있는 것이 보이지 않아서 코멘트를 남겨보았습니다!

RoadtoS7 replied:
오..! 그러네요..! 
[c52aec1](https://github.com/next-step/swift-blackjack/pull/12/commits/c52aec1a73752d9ff381dd50a66f34c5a1370329) 커밋에서 반영했습니다! 피드백 감사합니다!!

zdodev commented on Blackjack/WinLoseResult.swift:
```diff
@@ -0,0 +1,28 @@
+
+import Foundation
+
+struct WinLoseResult{
```
괄호가 붙어있습니다!

zdodev commented on Blackjack/WinLoseResult.swift:
```diff
@@ -0,0 +1,28 @@
+import Foundation
+
+struct WinLoseResult{
+    private let zero = 0
```
0을 나타내는 `zero` 프로퍼티를 선언해보셨네요!
0에 대한 상수를 나타내는 거라면 `Int.zero` 타입 프로퍼티를 사용해 보는 것은 어떨까요?

RoadtoS7 replied:
오 `Int.zero`라는 프로퍼티가 있군요..!! 피드백 감사합니다!!
[160edda](https://github.com/next-step/swift-blackjack/pull/12/commits/160eddaed851e8a9d41c3c9686651e34d07d5d79) 커밋에서 반영했습니다!

zdodev commented on BlackjackTest/BlackjackGameJudgeTest.swift:
```diff
@@ -0,0 +1,63 @@
+}
+
+class BlackjackGameJudgeTest: XCTestCase {
+    func test_winLoseResultScores_게임참가자들의_게임점수를_가지고_실패결과에_해당하는_WinLoseResults를_반환한다() {
```
메서드 이름의 `실패결과에 해당하는 WinLoseResults` 부분이 어떤 실패를 의미하는지 궁금합니다!

RoadtoS7 replied:
승패결과를 실패결과라고 잘못 적었습니다🤦‍♀️
[24511fd](https://github.com/next-step/swift-blackjack/pull/12/commits/24511fdafd84c7c09ca4447d8d54767315f72e4c) 커밋에서 반영했습니다! 피드백 감사합니다!

zdodev commented on BlackjackTest/BlackjackScoreTest.swift:
```diff
@@ -0,0 +1,64 @@
+        
+        // when
+        // then
+        XCTAssertFalse(blackjackScores.playerScores.contains(dealerScore))
+        XCTAssertTrue(blackjackScores.playerScores.contains(playerScore))
```
디미터의 법칙을 무조건 적용해야 한다는 건 어렵지 않을까 싶습니다! 특히 Swift 언어에서는 때에 따라 `.` 을 여러 번 사용하는 것이 더 좋을 때도 있다고 생각합니다!
기준을 가지고 `.` 을 여러 번 사용해서 가독성을 헤치지 않거나 오히려 좋아진다면 개인의 관점에 따라 사용해도 좋지 않을까 싶습니다.😁

zdodev commented on BlackjackTest/BlackjackScoreTest.swift:
```diff
@@ -0,0 +1,64 @@
+        XCTAssertTrue(blackjackScores.playerScores.contains(playerScore))
+    }
+    
+    func test_BlackjackSCores_dealerScore_딜러의_점수를_반환한다() {
```
오타가 있어보입니다!

RoadtoS7 replied:
[4cd173d](https://github.com/next-step/swift-blackjack/pull/12/commits/4cd173d61bfc0e3ca1fbebf959f07cc219305ea9)에서 수정했습니다! 피드백 감사합니다!!

zdodev reviewed (changes requested):
안녕하세요! @RoadtoS7 

블랙잭 2단계도 잘 구현해주셨습니다!
프로그래밍 요구 사항을 잘 지키고 구현하려고 하신 점이 잘 보였습니다!

몇 가지 코멘트 남겨보았습니다!
같이 의견 나눠보면 좋을 것 같습니다!😁

RoadtoS7 committed:
refactor: BlackjackGameJudge의 메서드 파라미터 명, 변수 명을 변경하여 가독성 개선

RoadtoS7 committed:
style: WinLoseResult에서 네이밍과 괄호 사이에 띄어쓰기 함

RoadtoS7 committed:
refactor: WinLoseResult에서 zero 상수 대신 Int.zero를 사용하도록 수정

RoadtoS7 committed:
refactor: winLoseResults()메서드에 대한 유닛테스트 이름 수정

- 수정 이유

승패 결과를 반환하는 것인데, 실패 결과를 반환하는 것이라고 오타가 났었다.

RoadtoS7 committed:
refactor: BlackjackScoreTest에서 dealerScore에 대한 유닛테스트 이름에 존재하는 오타 수정

RoadtoS7 committed:
style: main로직의 인덴테이션 맞춤

RoadtoS7 committed:
style: Card, AnswerParser 파일의 인덴테이션 맞춤

RoadtoS7 committed:
fix: NonDistributeCardDistributor에서 중복된 카드가 뽑히는 이슈 해결

- 이슈 원인
distribute()메서드에서 여러장의 카드를 출력할 때 (count 파라미터 값이 2이상인 경우), 뽑은 카드를 카드 집합에서 제거하지 않기 때문에
중복된 카드가 뽑힐 수 있다.

- 이슈 해결 방법
count 만큼 반복문을 돌며 카드를 뽑을 때, 카드를 뽑자마자 뽑은 카드를 카드 집합에서 제거하도록 했다.

RoadtoS7 committed:
refactor: BlackjackGameTest에서 오타 수정

RoadtoS7 committed:
refactor: 'test_BlackjackScores_dealerScore_딜러의_점수를_반환한다'에서 playerScore와 일치하지 않는지 테스트하도록 수정

zdodev commented on Blackjack/CardDistributor.swift:
```diff
@@ -31,12 +31,10 @@ class NonDuplicateCardDistributor: CardDistributor {
-        
-        pickedCards.forEach {
-            cards.remove($0)
+        let pickedCards = try (0..<count).map { _  -> Card in
+            let card = try pickCard()
+            cards.remove(card)
+            return card
         }
```
중복으로 카드가 뽑힐 수 있는 로직을 해결해보셨네요.👍

zdodev commented on Blackjack/WinLoseResult.swift:
```diff
@@ -0,0 +1,26 @@
+    let loseCount: Int
+    
+    init?(player: Player, winCount: Int, loseCount: Int) {
+        guard winCount >= Int.zero, loseCount >= Int.zero else { return nil }
```
간결한 것이 더 좋다면 `.zero` 도 괜찮아 보입니다!

RoadtoS7 replied:
[d5c9038](https://github.com/next-step/swift-blackjack/pull/12/commits/d5c903846e19ba0b2701ee6d0aa76896eaf81cd3) 커밋에서 반영했습니다! 피드백 감사합니다!

zdodev commented on BlackjackTest/BlackjackScoreTest.swift:
```diff
@@ -0,0 +1,65 @@
+        
+        // when
+        // then
+햐        XCTAssertEqual(blackjackScores.dealerScore, dealerScore)
```
에러가 발생하고 있어요!

RoadtoS7 replied:
[edbd9f3](https://github.com/next-step/swift-blackjack/pull/12/commits/edbd9f3a4a324886dd0b345bc041fb0af0c68376) 커밋에서 반영했습니다! 피드백 감사합니다!!

zdodev reviewed (approved):
안녕하세요! @RoadtoS7 

스텝2 단계도 많이 개선해주셨네요.👍
테스트를 수행할 수 없는 에러만 수정되면 바로 머지하도록 하겠습니다!

몇 가지 코멘트를 남겨보았습니다.
확인해보고 다음 스텝에서 고민해보면 좋을 것 같습니다! 고생하셨습니다!👏

RoadtoS7 committed:
refactor: Player에서 Hashable 제거, Equatable 채택하도록 수정

RoadtoS7 committed:
refactor: WinLoseResult에서 Int.zero -> .zero로 줄임

RoadtoS7 committed:
fix: 오타로 인한 컴파일 에러 수정

RoadtoS7 commented:
안녕하세요 리뷰어님! 수정 사항 확인 후 머지 부탁드립니다!!
